src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Collections.cs

```diff
@@ -101,40 +102,47 @@ internal sealed partial class ExpressionEval {
-                    case SetComprehension sc:
-                        var v2 = GetValueFromExpression(sc.Item) ?? UnknownType;
-                        return PythonCollectionType.CreateSet(Module, new[] { v2 });
-                    case DictionaryComprehension dc:
-                        var k = GetValueFromExpression(dc.Key) ?? UnknownType;
-                        var v = GetValueFromExpression(dc.Value) ?? UnknownType;
-                        return new PythonDictionary(new PythonDictionaryType(Interpreter.ModuleResolution.BuiltinsModule), new Dictionary<IMember, IMember> { { k, v } });
+        public IMember GetValueFromComprehensionInScope(Comprehension node, bool is3X = true) {
+            if (is3X) {
+                using (OpenScope(Module, node)) {
+                    ProcessComprehension(node);
+                    return GetValueFromComprehension(node);
                 }
+            }
 
-                return UnknownType;
-            } finally {
-                // Remove temporary variables since this is assignment and the right hand
-                // side comprehension does not leak internal variables into the scope.
-                var newVariables = CurrentScope.Variables.ToDictionary(k => k.Name, v => v);
-                var variables = (VariableCollection)CurrentScope.Variables;
-                foreach (var kvp in newVariables) {
-                    if (!oldVariables.ContainsKey(kvp.Key)) {
-                        variables.RemoveVariable(kvp.Key);
-                    }
+            ProcessComprehension(node);
+            return GetValueFromComprehension(node);
+        }
+
+        private IMember GetValueFromComprehension(Comprehension node) {
+            // TODO: Evaluate comprehensions to produce exact contents, if possible.
+            switch (node) {
+                case ListComprehension lc:
+                    var v1 = GetValueFromExpression(lc.Item) ?? UnknownType;
+                    return PythonCollectionType.CreateList(Module, new[] { v1 });
+                case SetComprehension sc:
+                    var v2 = GetValueFromExpression(sc.Item) ?? UnknownType;
+                    return PythonCollectionType.CreateSet(Module, new[] { v2 });
+                case DictionaryComprehension dc:
+                    var k = GetValueFromExpression(dc.Key) ?? UnknownType;
+                    var v = GetValueFromExpression(dc.Value) ?? UnknownType;
+                    return new PythonDictionary(new PythonDictionaryType(Interpreter.ModuleResolution.BuiltinsModule), new Dictionary<IMember, IMember> { { k, v } });
+            }
+            return UnknownType;
+        }
+
+        internal void ProcessComprehensionInScope(Comprehension node, bool is3X = true) {
+            // In 2X there is a bug where comprehension variables get leaked to outer scopes, fixed in 3X
+            if (is3X) {
+                using (OpenScope(Module, node)) {
```

Comment:
I can see why you did it this way, in order to have a conditional `using`. I'm not sure if we can do it in this project (due to the .NET standard 2.0 requirement), but maybe one of those newfangled C# 8 usings would work in a conditional, where you can say "this disposable gets disposed at return".

Otherwise, it might be okay to do something like this:

```csharp
var scope = Disposable.Empty;
if (is3x) {
    scope = OpenScope(module, node)
}

using (scope) {
    // something
}
```

Or rename the private `ProcessComprehension` function and make this one `ProcessComprehension`.

Reply:
Ok I've run into situations like this in the past, and at my old company, we renamed the function to `Function0` for example in this case we would have `ProcessComprehension` and `ProcessComprehension0`. Is that ok with our coding guidelines? Do you have a suggestion for a name?

Reply:
Not a fan of the `0` thing (not that I haven't seen it done). Maybe `ProcessComprehensionInScope` is actually a useful name, but reversed such that it's saying that "you're calling me in a scope".

Reply:
I might end up doing that nice trick with the disposable actually

Reply:
FYI you can probably shorten this function a little by writing a using statement like:

```csharp
using (is3x ? OpenScope(module, node) : Disposable.Empty) {
    // work
}
```

And prevent a future error of forgetting to change two calls.

Reply:
I think C# lets you pass a `null` disposable to a using and it'll skip over the dispose, but that feels very weird to me.